---
title: "CopyKAT: Rhabdomyosarcoma cell type tumor vs normal inference for `r {params$library_id}`"
output: 
  html_notebook:
  toc: true
toc_float: true
params:
  s3_singler_data_dir: "s3://nextflow-ccdl-results/scpca/processed/results/SCPCP000005" 
  s3_submitter_data_dir: "s3://sc-data-integration/scpca/celltype_sce"
  local_data_dir: "../data"
  library_id: "SCPCL000488"
  sample_id: "SCPCS000262"
  threads: 16
---
  
This notebook explores performance of the [`copyKat`](https://github.com/navinlabcode/copykat) package for calling tumor vs. normal cells in pediatric cancer scRNA-seq data.
Specifically, we look at a library from the `SCPCP000005` (RMS) project, where the submitter has annotated cell types which we can consider "ground truth" for the purposes of this exploration.
We'll assess agreement between the submitter tumor/normal classification and what `copyKat` infers as tumor/normal, across some different `copyKat` parameter values.

## Background: `copyKat` method

Reference: https://doi.org/10.1038/s41587-020-00795-2

To begin, here's a brief overview of the `copyKat` approach, parly quoting the paper throughout.
See Figure 1 in the paper for a method diagram:

* `copyKat` takes the _raw expression matrix_ and stabilizes variance, smooths the outliers
* A subset of cells is then identified to serve as a diploid baseline for subsequent identification of diploid/aneuploid
  * Cells are pooled into clusters with hierarchical clustering, and a GMM estimtes variance of each cluster
  * Cluster with minimal estimated variance is regarded as the "confident diploid cells"
  * **Caution:** This may fail if data has few normal cells and/or there are few CNA events.
  A "GMM definition mode" (sic; I think they meant "model"!) is offered to deal with this (described in paper bottom of pg 599), but implementation details to invoke this mode are thin. I _think_ it is automatically applied?
* MCMC is used to identify chromosome breakpoints with KS tests between gene windows
  * Significant test = probably a breakpoint
  * Final copy numbers for each window are the posterior means of all genes spanning breakpoints
  * These values are converted to genomic positions in 220-kb bins.
  _`copyKat` ultimately provides these values in the outputted CNA matrix._
* Hierarchical clustering identifies largest distance between aneuploid and diploid, but if distance not significant, they switch to the "GMM definition model" to predict tumor cells one-by-one.
* Finally, clustering delinates clonal substructure (not particularly relevant to us) for differential expression.


## Set Up

```{r}
suppressPackageStartupMessages({
  library(SingleCellExperiment)
  library(ggplot2)
  library(copykat) # https://github.com/navinlabcode/copykat
})
theme_set(theme_bw())

# define output path and SCE file
output_dir <-  here::here("celltype_annotation", 
                          "analysis", 
                          "copykat-results")
fs::dir_create(output_dir)

# Output file that compares results for different distance parameters
copykat_distance_sce <- file.path(output_dir,
                                  glue::glue("{params$library}_copykat_distance_sce.rds"))

# Output file that compares results for different kscut parameters
copykat_kscut_sce <- file.path(output_dir,
                                  glue::glue("{params$library}_copykat_kscut_sce.rds"))
```

The next chunk defines some functions for analyzing results. for preparing `copyKat` classification results for visualization and then making a barplot.


```{r}
# function to compare copykat results across values for _one_ varied parameter
#   with a barplot
compare_copykat_results_barplot <- function(copykat_df) {
  
  # prepare data frame for plotting
  copykat_df_long <- copykat_df |>
    tibble::rowid_to_column(var = "cell") |>
    tidyr::pivot_longer(
      contains("copykat"), 
      names_to = "copykat_parameter", 
      values_to = "prediction"
    ) |>
    dplyr::mutate(copykat_parameter = stringr::str_replace(copykat_parameter, 
                                                         "copykat_", ""))
  
  # make, return a barplot of predictions
  ggplot(copykat_df_long) + 
    aes(x = submitter_binary, fill = prediction) + 
    geom_bar(position = "fill",
             # use alpha to see background grid
             alpha = 0.75) + 
    facet_grid(vars(copykat_parameter)) +
    labs(
      x = "Submitter annotation",
      y = "Proportion of cells", 
      fill = "CopyKat prediction"
    ) + 
    # add guideline at y=0.5
    geom_hline(yintercept = 0.5, alpha = 0.25) 
}



# function to compare result tables both via overall counts and confusion matrix
compare_result_tables <- function(df, run_name) {
  # Takes data frame with three columns with fully defined values:
  ### `submitter`: full submitter annotations
  ### `submitter_binary`: submitter annotations reduced to tumor/normal only
  #### `copykat`: copykat annotations, considering only diploid/aneuploid
  # Returns overall table comparison and result from caret::confusionMatrix()

  print(glue::glue("=================== {run_name} ======================"))
  
  # Result 1: Table comparing full results
  print(
    table(df$submitter, df$copykat)
  )
  
  cat("\n ===================================\n ")
  # Result 2: confusion matrix
  
  # remove "not.defined" from copyKat and update levels
  df_caret <- df |> 
    dplyr::filter(copykat != "not.defined") |>
    # copykat values need to match submitter_binary values
    dplyr::mutate(copykat = ifelse(
      copykat == "diploid", "Normal", "Tumor"
    )) |>
    # Tumor should be first as the "positive" value
    dplyr::mutate(
      copykat = forcats::fct_relevel(copykat, "Tumor"),
      submitter_binary = forcats::fct_relevel(submitter_binary, "Tumor")
    )
  
  print(
    caret::confusionMatrix(
      table(df_caret$submitter_binary, 
            df_caret$copykat)
    )
  )
}
```


In the next chunk, we obtain the _submitter-annotated_ SCE from S3, if it is not already locally present.
We'll wait to read it in when actually running `copyKat`, mostly for memory reasons.


```{r}
# make local data directory if it doesn't exist
if (!dir.exists(params$local_data_dir)) {
  dir.create(params$local_data_dir, recursive = TRUE)
}

# build path to submitter annotated sce file 
submitter_annotated_sce_file <- glue::glue("{params$library_id}_processed_celltype.rds")


local_submitter_sce_path <- file.path(params$local_data_dir,
                                      params$sample_id,
                                      submitter_annotated_sce_file)

# get submitter sce from S3 if not already present locally
if (!(file.exists(local_submitter_sce_path))) {
  sync_call <- glue::glue('op run -- aws s3 sync {params$s3_submitter_data_dir} {params$local_data_dir}/{params$sample_id} --exclude "*" --include "{submitter_annotated_sce_file}"')
  system(sync_call, ignore.stdout = TRUE) 
}
```


## Compare CopyKAT distance parameters


### Run copyKat

The `copyKat` [GitHub `README` file](https://github.com/navinlabcode/copykat/blob/master/README.md) includes this caveat about the `"distance"` parameter:

> One struggling observation is that none of one clustering method could fit all datasets. In this version, I add a distance parameters for clustering that include "euclidean" distance and correlational distance, ie. 1-"pearson" and "spearman" similarity. In general, corretional distances tend to favor noisy data, while euclidean distance tends to favor data with larger CN segments.

Therefore we'll begin by comparing `"euclidian"`, `"spearman"`, and `"pearson"` values for [this parameter](https://github.com/navinlabcode/copykat/blob/b795ff793522499f814f6ae282aad1aab790902f/R/copykat.R#L251-L255).


```{r}
# Only run copykat if the final SCE does not exist
if (file.exists(copykat_distance_sce)) {
  
  sce <- readr::read_rds(copykat_distance_sce)
  
} else {
  
  # Otherwise, read in initial SCE and run copyKat
  sce <- readr::read_rds(local_submitter_sce_path)
    
  purrr::walk(
    c("spearman", "pearson", "euclidean"), 
    \(distance_parameter) {
      
      copykat_result <- copykat(
        rawmat = as.matrix(counts(sce)), 
        id.type = "E", # we have Ensembl gene ids, not the default gene symbols ("S")
        sam.name = glue::glue("{params$library_id}-{distance_parameter}"), # unique name for each 
        distance = distance_parameter, 
        plot.genes = FALSE,
        n.cores = params$threads
      )
      
      # save prediction to sce
      colname <- glue::glue("copykat_{distance_parameter}")
      colData(sce)[[colname]] <- copykat_result$prediction$copykat.pred
      
      # export full result
      output_file <- file.path(output_dir,
                              glue::glue("{params$library}_copykat_{distance_parameter}.rds"))
      readr::write_rds(copykat_result, output_file)
    }
  )
  
  # finally, export SCE with all copykat predictions
  readr::write_rds(sce, copykat_distance_sce)
} 
```

### Comparing results


We'll prepare the data for comparison and begin with a barplot of tumor/normal classifications across `copyKat` parameters:

```{r}
# Prepare data for comparisons
copykat_df <- colData(sce) |>
  as.data.frame() |>
  # better printing 
  tibble::as_tibble() |>
  dplyr::select(submitter = celltype, contains("copykat")) |>
  # remove NA submitter annotations
  tidyr::drop_na(submitter) |>
  dplyr::mutate(submitter_binary = ifelse(
    stringr::str_starts(submitter, "Tumor"), 
    "Tumor", 
    "Normal"
  )) 

# Plot the data:
compare_copykat_results_barplot(copykat_df) + 
  ggtitle("Comparing distance parameters")
```

Let's also look at the binary submitter celltypes to get a sense of balance:

```{r}
table(copykat_df$submitter_binary)
```


Next, we'll look at tables for correspondence between submitter cell type annotations and `copyKat` inferences, as well as confusion matrix results for each comparison.
For `copyKat`, "diploid" is normal and "aneuploid" is tumor.

```{r}
copykat_df |>
  dplyr::select(submitter, submitter_binary, copykat = copykat_euclidian) |>
  compare_result_tables("euclidian distance") 

copykat_df |>
  dplyr::select(submitter, submitter_binary, copykat = copykat_pearson) |>
  compare_result_tables("pearson distance") 

copykat_df |>
  dplyr::select(submitter, submitter_binary, copykat = copykat_spearman) |>
  compare_result_tables("spearman distance") 

```

## Compare CopyKAT KS.cut parameters


### Run copyKat

In the previous section, we saw that correlational distance is likely to be better than euclidian.
In this section, we'll again run `copyKat`, **but this time using _only_ the "pearson" distance**, at a variety of `KS.cut` parameters.

As described in the [GitHub `README` file](https://github.com/navinlabcode/copykat/blob/master/README.md), 

> KS.cut is the segmentation parameter, ranging from 0 to 1. Increasing KS.cut decreases sensitivity, i.e. less segments/breakpoints. Usually it works in a range of 0.05-0.15.

For this, we'll start with the SCE from the last section, and build it up from there with other `KS.cut` values.
Since it was already run with default `KS.cut = 0.1`, we'll test a few additional values here and compare those to above.

```{r}
if (file.exists(copykat_kscut_sce)) {
  # if the SCE exists, read it in
  sce <- readr::read_rds(copykat_kscut_sce)

} else if (file.exists(copykat_distance_sce)) {
  
  # if only the _distance_ SCE exists, read in as a starting point
  sce <- readr::read_rds(copykat_distance_sce)
  
  # clean up columns for this analysis
  sce$copykat_0.1 <- sce$copykat_pearson # no need to redo this one! 
  sce$copykat_pearson  <- NULL
  sce$copykat_spearman <- NULL
  sce$copykat_euclidian <- NULL
  
  # and run!
  purrr::walk(
    seq(0.025, 0.15, 0.025)[-4], # can skip 0.1
    \(kscut_value) {
      
      copykat_result <- copykat(
        rawmat = as.matrix(counts(sce)), 
        id.type = "E", # we have Ensembl gene ids, not the default gene symbols ("S")
        sam.name = glue::glue("{params$library_id}-{kscut_value}"), # unique name for each 
        distance = "pearson", 
        KS.cut = kscut_value,
        plot.genes = FALSE,
        n.cores = params$threads
      )
      
      # save prediction to sce
      colname <- glue::glue("copykat_{kscut_value}")
      colData(sce)[[colname]] <- copykat_result$prediction$copykat.pred
      
      # export full result
      output_file <- file.path(output_dir,
                              glue::glue("{params$library}_copykat_{kscut_value}.rds"))
      readr::write_rds(copykat_result, output_file)
    }
  )
  
  # finally, export SCE with all copykat kscut predictions
  readr::write_rds(sce, copykat_kscut_sce)
  
} else {
  stop("Error: At least one of those SCE files should have existed by now!")
}
```


### Comparing results

As before, we'll prepare the data for comparison and begin with a barplot of tumor/normal classifications across `copyKat` parameters:

```{r, fig.height = 5, fig.width = 4}
# Prepare data for comparisons
copykat_df <- colData(sce) |>
  as.data.frame() |>
  # better printing 
  tibble::as_tibble() |>
  dplyr::select(submitter = celltype, contains("copykat")) |>
  # remove NA submitter annotations
  tidyr::drop_na(submitter) |>
  dplyr::mutate(submitter_binary = ifelse(
    stringr::str_starts(submitter, "Tumor"), 
    "Tumor", 
    "Normal"
  )) 

# Plot the data:
compare_copykat_results_barplot(copykat_df) + 
  ggtitle("Comparing KS.cut parameters")
```


Next, we'll look at tables for correspondence between submitter cell type annotations and `copyKat` inferences, as well as confusion matrix results for each comparison.
For `copyKat`, "diploid" is normal and "aneuploid" is tumor.

```{r}
purrr::walk(
  glue::glue("copykat_{seq(0.025, 0.15, 0.025)}"), 
  \(column_name) {
    
  copykat_df |>
    dplyr::select(submitter, submitter_binary, copykat = as.symbol(column_name) ) |>
    compare_result_tables(column_name) 
  }
)
```



## Conclusions

`copyKat` has the best chance of success with one of the correlational distances (spearman or pearson), and the default `KS.cut` parameter of 0.1 may indeed be a reasonable default for us (although many of the results across this parameter are similar).

It is also worth noting some drawbacks of the `copyKat` package:

* `copyKat` [hardcodes a random seed](https://github.com/navinlabcode/copykat/blob/b795ff793522499f814f6ae282aad1aab790902f/R/copykat.R#L29-L32) so we can't readily assess bias from starting conditions.
* It emits _a lot of output_, and we don't have much control over it! 
  * This includes both `print()` statements and output files, some of which are optional but not all; and it seems the optional ones may still be exported even if you turn them off?
  * The _lots and lots of output_ is not seen in this notebook because it was rendered after files already existed; the output _cannot_ be suppressed by turning off warnings or messages, because they are all print statements. 
  * Of this output, we get a `"cell: #"` printed for each cell, and we also see that the majority also have a printed statement: `"WARNING! NOT CONVERGENT!"`.
  This latter statement is _printed_ from a [dependency](https://github.com/cran/mixtools/blob/4af1e2789bcea7df3c1775a53cd05b37ec3185d0/R/normalmixEM.R#L129-L131) during the `copyKat` step to identify a baseline diploid cell population.
* There is no clear way to assess confidence from `copyKat` output.
  * In the `copyKay` paper, they performed bootstrapping to assess performance, which may be something we'll have to do to get some kind of confidence measure. 
  From the paper, " Bootstrapping was performed to resample single cells and estimate the variation in detection sensitivity. On average, we estimated that 19% of CNAs were detected at 1-Mb resolution, 56% were detected at 5-Mb resolution and 88% were detected at 20-Mb resolution, which agrees with our theoretical calculation of a 5-Mb average genomic resolution."


Based on these drawbacks, if we want to use `copyKat` moving forward, we might consider forking the repository and modifying aspects that best suit our needs:

* More control over the seed
* Less printing and fewer output files, where possible


## Session Info

```{r}
sessionInfo()
```
